File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> set;
        int left = 0;
        int right = 0;
        int ans = 0;
        while(right < s.size()) {
            if(!set.count(s[right])) {
                ans = max(ans, right-left+1);
                set.insert(s[right++]);
            }
            else {
                set.erase(s[left++]);
            }
        }
        return ans;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s.length() == 0) return 0;
        int max = Integer.MIN_VALUE;
        int i = 0, j = 0;
        HashSet<Character> set = new HashSet();
        while(i < s.length()) {
            if(!set.contains(s.charAt(i))) {
                set.add(s.charAt(i));
                max = Math.max(max, i-j+1);
                i++;
            }
            else {
                set.remove(s.charAt(j));
                j++;
            }
        }
        return max;
    }
}